### PR TITLE
Offstrategy keyspace compaction task

### PR DESCRIPTION
**Pull request by Deexie**
Task manager task implementations of classes that cover
offstrategy keyspace compaction which can be start through
/storage_service/keyspace_compaction/ api.

Top level task covers the whole compaction and creates child
tasks on each shard.

### CODE REVIEW

**Comment by Deexie** [comment]
Draft until #12661 gets merged.

File: compaction/compaction_task.hh [hunk 1]
```diff
@@ -10,6 +10,28 @@
 
 #include "tasks/task_manager.hh"
```
Comment (denesb) [inline compaction/compaction_task.hh:12]: typo in subject: s/compacition/compaction/

File: compaction/compaction_task.hh [hunk 2]
```diff
@@ -10,6 +10,28 @@
...
 
+class compaction_task_impl : public tasks::task_manager::task::impl {
+public:
+    compaction_task_impl(tasks::task_manager::module_ptr module,
+            tasks::task_id id,
+            unsigned sequence_number,
+            std::string keyspace,
+            std::string table,
+            std::string entity,
```
Comment (denesb) [inline compaction/compaction_task.hh:22]: All current implementations pass empty string for this `entity` parameter. Is that really needed? Will there be a future task type which passes something here?

**Review by denesb — COMMENTED** [review]
AFAIK compactions do support aborting, but I don't see abort implemented in this PR. Will that come later, or it is not planned?

**Comment by Deexie** [comment]
- rewrite

**Comment by Deexie** [comment]
@bhalevy review please

**Comment by Deexie** [comment]
- rebase

**Review by denesb — APPROVED** [review]
@bhalevy please also review.

File: compaction/task_manager_module.hh [hunk 3]
```diff
@@ -149,6 +149,70 @@ protected:
...
+    }
+protected:
+    virtual future<> run() override = 0;
+};
+
+class offstrategy_keyspace_compaction_task_impl : public offstrategy_compaction_task_impl {
+private:
+    sharded<replica::database>& _db;
+    std::vector<table_id> _table_infos;
+    bool& _res;
```
Comment (bhalevy) [inline compaction/task_manager_module.hh:177]: I think `_needed` would be better than `_res` like in `shard_offstrategy_keyspace_compaction_task_impl` below

**Review by bhalevy — COMMENTED** [review]
LGTM overall. I just had a couple suggestions

File: compaction/task_manager_module.hh [hunk 4]
```diff
@@ -149,6 +149,70 @@ protected:
...
+        , _db(db)
+        , _table_infos(std::move(table_infos))
+        , _res(res)
+    {}
+protected:
+    virtual future<> run() override;
+};
+
+class shard_offstrategy_keyspace_compaction_task_impl : public offstrategy_compaction_task_impl {
+private:
```
Comment (raphaelsc) [inline compaction/task_manager_module.hh:194]: shard means shard local right? perhaps local would have been a better prefix. it's not immediately clear to me what shard means, unless  i peek into the implementation.

**Comment by Deexie** [comment]
- _res -> _needed
- shard_offstrategy_keyspace_compaction_task_impl -> local_offstrategy_keyspace_compaction_task_impl
- capture this in local_offstrategy_keyspace_compaction_task_impl::run()

**Review by bhalevy — APPROVED** [review]

**Review by raphaelsc — APPROVED** [review]

**Pull request merged**